src/Microsoft.OData.Client/BulkUpdateSaveResult.cs

```diff
@@ -202,19 +307,370 @@ protected override ODataRequestMessageWrapper CreateRequestMessage(string method
+        /// <param name="operationResponseTracker">Tracks the operation responses for the operations that have been performed.</param>
+        private void HandleInnerBulkUpdateResponse(
+            List<IMaterializerState> entries, 
+            bool isTopLevelResponse, 
+            Stack<(int lastVisited, IReadOnlyList<Descriptor> descriptors)> adjacentDescriptors,
+            Stack<List<OperationResponse>> operationResponseTracker) 
+        {  
+            foreach (IMaterializerState entry in entries)
+            {
+                if (entry is MaterializerEntry materializerEntry)
+                {
+                    CreateOperationResponse(materializerEntry, isTopLevelResponse, adjacentDescriptors, operationResponseTracker);
+                   
+                    foreach (IMaterializerState nestedItem in materializerEntry?.NestedItems)
+                    {
+                        if (nestedItem is MaterializerNestedEntry materializerNestedEntry)
+                        {
+                            foreach (IMaterializerState nestedEntries in materializerNestedEntry?.NestedItems)
+                            {
+                                if (nestedEntries is MaterializerDeltaFeed feed)
+                                {
+                                    this.HandleInnerBulkUpdateResponse(feed.Entries, false, adjacentDescriptors, operationResponseTracker);
+                                }
+                            }
+                        }
+                    }
+                }
+                else if (entry is MaterializerDeletedEntry deletedEntry)
+                {
+                    CreateOperationResponse(deletedEntry, isTopLevelResponse, adjacentDescriptors, operationResponseTracker);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Processes the bulk update <see cref="IODataResponseMessage"/> response and caches the result.
+        /// </summary>
+        /// <param name="response">An instance of the <see cref="IODataResponseMessage"/> for the bulk update response.</param>
+        /// <param name="responseStream">The response stream.</param>
+        private void HandleBulkUpdateResponse(IODataResponseMessage response, Stream responseStream)
+        {
+            EntityDescriptor entityDescriptor = this.bulkUpdateGraph.TopLevelDescriptors[0];
+            MaterializerDeltaFeed deltaFeed = default;
+
+            // We are setting the MaxProtocolVersion to v4.01 when reading bulk update responses.
+            this.RequestInfo.Context.MaxProtocolVersion = ODataProtocolVersion.V401;
+            Exception exception = HandleResponse(
+                this.RequestInfo, 
+                (HttpStatusCode)response.StatusCode, 
+                response.GetHeader(XmlConstants.HttpODataVersion), 
+                () => { return ResponseStream; },

```

Comment:
Did you consider the suggestion by @habbes here? https://github.com/OData/odata.net/pull/2627/files#r1164080720